# Importing libraries, df_ords_prods_merge & checking outcome

In [2]:
import pandas as pd
import numpy as np
import os

In [4]:
path=r'/Users/stefaniadaskalaki/Data Immersion/4 Achievement Python/CF4.2 Instacart Basket Analysis'

In [5]:
df_ords_prods_merge=pd.read_pickle(os.path.join(path, '02. Data', '02. Prepared Data', 'ords_prods_merge.pkl'))

In [6]:
df_ords_prods_merge.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_prior_order  new_customer product_id  add_to_cart_order  \
0                     NaN          True        196                  1   
1                    15.0         False        196                  1   
2                    21.0         False        196                  1   
3                    29.0         False        196                  1   
4                    28.0         False        196                  1   

   reordered  ... aisle_id department_id prices  _merge    price_range_loc  \
0          0  ...       77             7    9.0    both  Mid-range product   
1          1  ...       77             7    9.0    both  Mid-range product   
2          1  ...       77             7    9.0    both  Mid-range product   
3          1  ...       77             7    9.0    both  Mid-range product   
4          1  ...       77             7    9.0    both  Mid-range product   

      Busiest_day    Busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy    Slowest days        Average orders        10   
2  Regularly busy    Slowest days           Most orders        10   
3      Least busy    Slowest days        Average orders        10   
4      Least busy    Slowest days           Most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 21 columns]

In [8]:
df_ords_prods_merge.shape

(32404859, 21)

# 2. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [9]:
df_ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
2     17.277920
20    16.473447
21    22.902379
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
Name: order_number, dtype: float64

# 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code

It is interesting to see that certain depts haven't changed almost at all such as dept 7 (only 0.2) and the rest have changed from 1-4 points. The change is of course expected since the first time we run this code was only in the first 1000000 rows instead now is the full dataset.

# 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions

In [10]:
df_ords_prods_merge['max_order']=df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [11]:
df_ords_prods_merge.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   
4   431534       1             5                  4                 15   

   days_since_prior_order  new_customer product_id  add_to_cart_order  \
0                     NaN          True        196                  1   
1                    15.0         False        196                  1   
2                    21.0         False        196                  1   
3                    29.0         False        196                  1   
4                    28.0         False        196                  1   

   reordered  ... aisle_id department_id prices  _merge    price_range_loc  \
0          0  ...       77             7    9.0    both  Mid-range product   
1          1  ...       77             7    9.0    both  Mid-range product   
2          1  ...       77             7    9.0    both  Mid-range product   
3          1  ...       77             7    9.0    both  Mid-range product   
4          1  ...       77             7    9.0    both  Mid-range product   

      Busiest_day    Busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy        Average orders        10   
1  Regularly busy    Slowest days        Average orders        10   
2  Regularly busy    Slowest days           Most orders        10   
3      Least busy    Slowest days        Average orders        10   
4      Least busy    Slowest days           Most orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 21 columns]

In [12]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag']='Loyal customer'

In [13]:
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order']<=40)&(df_ords_prods_merge['max_order']>10),'loyalty_flag']='Regular customer'

In [14]:
df_ords_prods_merge.loc[df_ords_prods_merge['max_order']<=10,'loyalty_flag']='New customer'

In [15]:
df_ords_prods_merge['loyalty_flag'].value_counts()

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [16]:
df_ords_prods_merge['loyalty_flag']

0               New customer
1               New customer
2               New customer
3               New customer
4               New customer
                  ...       
32404854    Regular customer
32404855    Regular customer
32404856        New customer
32404857        New customer
32404858    Regular customer
Name: loyalty_flag, Length: 32404859, dtype: object

# 5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers

In [17]:
df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'sum']})

prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2

what is interesting in the results is that the Loyal customers are actually purchasing the cheapest products and not bringing the highest category revenue. However the regular customers are the top performers in terms of revenue. Even though they are buying high value products there is a mix with also lower priced products. However they are the ones bringing the most revenue with 85.74% more than the Loyal customers and 138.99% more than the New customers. Nevertheless the New customers are tending to purchase the highest priced products with avg price 13.3 euros. This can signify that they are only coming whenever they need these speciality items that can't find in their regular shop or that it was random occasions and the wanted the best option.

# 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

## I need to perform the following steps in order to create a new column with the avg_price for each customer
 1.Split the data into groups based on the “user_id” column
 2.Apply the transform() function on the “prices” column to generate the maximum orders for each user
 3.Create a new column , “avg_price” into which I’ll place the results of the aggregation

In [18]:
df_ords_prods_merge['avg_price']=df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.max)

In [20]:
df_ords_prods_merge.head(20)

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  2968173      15            15                  1                  9   
11  1870022      15            17                  2                 16   
12  1911383      15            18                  2                 11   
13  2715276      15            21                  1                  9   
14   487368      15            22                  1                 10   
15  2293453      19             2                  5                 14   
16  1973799      19             5                  6                 12   
17   532817      19             7                  4                 17   
18  1573906      21            10                  3                 10   
19  1593000      31            10                  3                  8   

    days_since_prior_order  new_customer product_id  add_to_cart_order  \
0                      NaN          True        196                  1   
1                     15.0         False        196                  1   
2                     21.0         False        196                  1   
3                     29.0         False        196                  1   
4                     28.0         False        196                  1   
5                     19.0         False        196                  1   
6                     20.0         False        196                  1   
7                     14.0         False        196                  2   
8                      0.0         False        196                  4   
9                     30.0         False        196                  1   
10                     7.0         False        196                  2   
11                     8.0         False        196                  6   
12                     7.0         False        196                  1   
13                     7.0         False        196                  1   
14                    14.0         False        196                  1   
15                     6.0         False        196                  3   
16                     8.0         False        196                 15   
17                     6.0         False        196                  1   
18                     6.0         False        196                  2   
19                     7.0         False        196                 17   

    reordered  ... department_id prices _merge    price_range_loc  \
0           0  ...             7    9.0   both  Mid-range product   
1           1  ...             7    9.0   both  Mid-range product   
2           1  ...             7    9.0   both  Mid-range product   
3           1  ...             7    9.0   both  Mid-range product   
4           1  ...             7    9.0   both  Mid-range product   
5           1  ...             7    9.0   both  Mid-range product   
6           1  ...             7    9.0   both  Mid-range product   
7           1  ...             7    9.0   both  Mid-range product   
8           1  ...             7    9.0   both  Mid-range product   
9           1  ...             7    9.0   both  Mid-range product   
10          0  ...             7    9.0   both  Mid-range product   
11          1  ...             7    9.0   bo

## Deriving spending_flag column with loc function and checking outcome

In [21]:
df_ords_prods_merge.loc[df_ords_prods_merge['avg_price'] >= 10, 'spending_flag']='High spender'

In [22]:
df_ords_prods_merge.loc[df_ords_prods_merge['avg_price'] < 10, 'spending_flag']='Low spender'

In [24]:
df_ords_prods_merge['spending_flag'].value_counts()

High spender    32372969
Low spender        31890
Name: spending_flag, dtype: int64

In [25]:
df_ords_prods_merge.head(20)

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  2968173      15            15                  1                  9   
11  1870022      15            17                  2                 16   
12  1911383      15            18                  2                 11   
13  2715276      15            21                  1                  9   
14   487368      15            22                  1                 10   
15  2293453      19             2                  5                 14   
16  1973799      19             5                  6                 12   
17   532817      19             7                  4                 17   
18  1573906      21            10                  3                 10   
19  1593000      31            10                  3                  8   

    days_since_prior_order  new_customer product_id  add_to_cart_order  \
0                      NaN          True        196                  1   
1                     15.0         False        196                  1   
2                     21.0         False        196                  1   
3                     29.0         False        196                  1   
4                     28.0         False        196                  1   
5                     19.0         False        196                  1   
6                     20.0         False        196                  1   
7                     14.0         False        196                  2   
8                      0.0         False        196                  4   
9                     30.0         False        196                  1   
10                     7.0         False        196                  2   
11                     8.0         False        196                  6   
12                     7.0         False        196                  1   
13                     7.0         False        196                  1   
14                    14.0         False        196                  1   
15                     6.0         False        196                  3   
16                     8.0         False        196                 15   
17                     6.0         False        196                  1   
18                     6.0         False        196                  2   
19                     7.0         False        196                 17   

    reordered  ... prices _merge    price_range_loc     Busiest_day  \
0           0  ...    9.0   both  Mid-range product  Regularly busy   
1           1  ...    9.0   both  Mid-range product  Regularly busy   
2           1  ...    9.0   both  Mid-range product  Regularly busy   
3           1  ...    9.0   both  Mid-range product      Least busy   
4           1  ...    9.0   both  Mid-range product      Least busy   
5           1  ...    9.0   both  Mid-range product  Regularly busy   
6           1  ...    9.0   both  Mid-range product  Regularly busy   
7           1  ...    9.0   both  Mid-range product  Regularly busy   
8           1  ...    9.0   both  Mid-range product  Regularly busy   
9           1  ...    9.0   both  Mid-range product      Least busy   
10          0  ...    9.0   both  Mid-range product  Regularly busy   
11          1  ...  

# 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [26]:
df_ords_prods_merge['frequency_order_median']=df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.max)

In [28]:
df_ords_prods_merge['frequency_order_median']

0           30.0
1           30.0
2           30.0
3           30.0
4           30.0
            ... 
32404854    30.0
32404855    30.0
32404856    30.0
32404857    30.0
32404858    30.0
Name: frequency_order_median, Length: 32404859, dtype: float64

In [27]:
df_ords_prods_merge.head(20)

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  2968173      15            15                  1                  9   
11  1870022      15            17                  2                 16   
12  1911383      15            18                  2                 11   
13  2715276      15            21                  1                  9   
14   487368      15            22                  1                 10   
15  2293453      19             2                  5                 14   
16  1973799      19             5                  6                 12   
17   532817      19             7                  4                 17   
18  1573906      21            10                  3                 10   
19  1593000      31            10                  3                  8   

    days_since_prior_order  new_customer product_id  add_to_cart_order  \
0                      NaN          True        196                  1   
1                     15.0         False        196                  1   
2                     21.0         False        196                  1   
3                     29.0         False        196                  1   
4                     28.0         False        196                  1   
5                     19.0         False        196                  1   
6                     20.0         False        196                  1   
7                     14.0         False        196                  2   
8                      0.0         False        196                  4   
9                     30.0         False        196                  1   
10                     7.0         False        196                  2   
11                     8.0         False        196                  6   
12                     7.0         False        196                  1   
13                     7.0         False        196                  1   
14                    14.0         False        196                  1   
15                     6.0         False        196                  3   
16                     8.0         False        196                 15   
17                     6.0         False        196                  1   
18                     6.0         False        196                  2   
19                     7.0         False        196                 17   

    reordered  ... _merge    price_range_loc     Busiest_day    Busiest_days  \
0           0  ...   both  Mid-range product  Regularly busy  Regularly busy   
1           1  ...   both  Mid-range product  Regularly busy    Slowest days   
2           1  ...   both  Mid-range product  Regularly busy    Slowest days   
3           1  ...   both  Mid-range product      Least busy    Slowest days   
4           1  ...   both  Mid-range product      Least busy    Slowest days   
5           1  ...   both  Mid-range product  Regularly busy  Regularly busy   
6           1  ...   both  Mid-range product  Regularly busy    Busiest days   
7           1  ...   both  Mid-range product  Regularly busy    Busiest days   
8           1  ...   both  Mid-range product  Regularly busy    Busiest days   
9           1  ...   both  Mid-range product      Least busy    Slowest 

## Deriving frequency_order_median_flag column with loc function and checking outcome


In [29]:
df_ords_prods_merge.loc[df_ords_prods_merge['frequency_order_median'] > 20, 'frequency_order_median_flag']='Non-frequent customer'

In [30]:
df_ords_prods_merge.loc[(df_ords_prods_merge['frequency_order_median'] > 10)&(df_ords_prods_merge['frequency_order_median'] <= 20), 'frequency_order_median_flag']='Regular customer'

In [31]:
df_ords_prods_merge.loc[df_ords_prods_merge['frequency_order_median'] <= 10, 'frequency_order_median_flag']='Frequent customer'

In [32]:
df_ords_prods_merge['frequency_order_median_flag'].value_counts()

Non-frequent customer    24487186
Regular customer          6533511
Frequent customer         1384157
Name: frequency_order_median_flag, dtype: int64

In [34]:
df_ords_prods_merge.head(20)

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  2968173      15            15                  1                  9   
11  1870022      15            17                  2                 16   
12  1911383      15            18                  2                 11   
13  2715276      15            21                  1                  9   
14   487368      15            22                  1                 10   
15  2293453      19             2                  5                 14   
16  1973799      19             5                  6                 12   
17   532817      19             7                  4                 17   
18  1573906      21            10                  3                 10   
19  1593000      31            10                  3                  8   

    days_since_prior_order  new_customer product_id  add_to_cart_order  \
0                      NaN          True        196                  1   
1                     15.0         False        196                  1   
2                     21.0         False        196                  1   
3                     29.0         False        196                  1   
4                     28.0         False        196                  1   
5                     19.0         False        196                  1   
6                     20.0         False        196                  1   
7                     14.0         False        196                  2   
8                      0.0         False        196                  4   
9                     30.0         False        196                  1   
10                     7.0         False        196                  2   
11                     8.0         False        196                  6   
12                     7.0         False        196                  1   
13                     7.0         False        196                  1   
14                    14.0         False        196                  1   
15                     6.0         False        196                  3   
16                     8.0         False        196                 15   
17                     6.0         False        196                  1   
18                     6.0         False        196                  2   
19                     7.0         False        196                 17   

    reordered  ...    price_range_loc     Busiest_day    Busiest_days  \
0           0  ...  Mid-range product  Regularly busy  Regularly busy   
1           1  ...  Mid-range product  Regularly busy    Slowest days   
2           1  ...  Mid-range product  Regularly busy    Slowest days   
3           1  ...  Mid-range product      Least busy    Slowest days   
4           1  ...  Mid-range product      Least busy    Slowest days   
5           1  ...  Mid-range product  Regularly busy  Regularly busy   
6           1  ...  Mid-range product  Regularly busy    Busiest days   
7           1  ...  Mid-range product  Regularly busy    Busiest days   
8           1  ...  Mid-range product  Regularly busy    Busiest days   
9           1  ...  Mid-range product      Least busy    Slowest days   
10          0  ...  Mid-range product  Regularly busy    Busiest days

# Exporting df_ords_prods_merge to pickle

In [38]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02. Data','02. Prepared Data', 'ords_prods_merge48.pkl'))